sycl/source/detail/global_handler.cpp

```diff
@@ -0,0 +1,137 @@
+#include <detail/platform_impl.hpp>
+#include <detail/plugin.hpp>
+#include <detail/program_manager/program_manager.hpp>
+#include <detail/scheduler/scheduler.hpp>
+
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+GlobalHandler *SyclGlobalObjectsHandler;
+// SpinLock is chosen because, unlike std::mutex, it can be zero initialized,
+// which spares us from dealing with global constructor/destructor call order.
+SpinLock GlobalWritesAllowed;
+SpinLock ShutdownLock;
+
+GlobalHandler::GlobalHandler() = default;
+GlobalHandler::~GlobalHandler() = default;
+
+GlobalHandler &GlobalHandler::instance() {
+  if (!SyclGlobalObjectsHandler) {
+    const std::lock_guard<SpinLock> Lock{GlobalWritesAllowed};
+    if (!SyclGlobalObjectsHandler) {
+      SyclGlobalObjectsHandler = new GlobalHandler();
+    }
+  }
+
+  return *SyclGlobalObjectsHandler;
+}
+
+Scheduler &GlobalHandler::getScheduler() {
+  if (!MScheduler) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MScheduler) {
+      MScheduler = std::make_unique<Scheduler>();
+    }
+  }
+  return *MScheduler;
+}
+ProgramManager &GlobalHandler::getProgramManager() {
+  if (!MProgramManager) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MProgramManager) {
+      MProgramManager = std::make_unique<ProgramManager>();
+    }
+  }
+  return *MProgramManager;
+}
+Sync &GlobalHandler::getSync() {
+  if (!MSync) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MSync) {
+      MSync = std::make_unique<Sync>();
+    }
+  }
+  return *MSync;
+}
+std::vector<PlatformImplPtr> &GlobalHandler::getPlatformCache() {
+  if (!MPlatformCache) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlatformCache) {
+      MPlatformCache = std::make_unique<std::vector<PlatformImplPtr>>();
+    }
+  }
+  return *MPlatformCache;
+}
+std::mutex &GlobalHandler::getPlatformMapMutex() {
+  if (!MPlatformMapMutex) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlatformMapMutex) {
+      MPlatformMapMutex = std::make_unique<std::mutex>();
+    }
+  }
+  return *MPlatformMapMutex;
+}
+std::mutex &GlobalHandler::getFilterMutex() {
+  if (!MFilterMutex) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MFilterMutex) {
+      MFilterMutex = std::make_unique<std::mutex>();
+    }
+  }
+  return *MFilterMutex;
+}
+std::vector<plugin> &GlobalHandler::getPlugins() {
+  if (!MPlugins) {
+    const std::lock_guard<SpinLock> Lock{MFieldsLock};
+    if (!MPlugins) {
+      MPlugins = std::make_unique<std::vector<plugin>>();
+    }
+  }
+  return *MPlugins;
+}
+
+void shutdown() {
+  if (SyclGlobalObjectsHandler) {
+    const std::lock_guard<SpinLock> Lock{ShutdownLock};
+    if (SyclGlobalObjectsHandler) {
+      // Acquire fields lock to make sure no thread creates fields while
+      // object destruction is in flight.
+      const std::lock_guard<SpinLock> Lock{
```

Comment:
Not sure this helps, please, consider the case:
1. Thread 1 is on line 96.
2. Thread 2 is completes shutdown function.
3. Thread 1 tries to lock `MFieldsLock` and crashes because this is already destroyed.

Not sure how to address this issue so far, I will think an comment.

Reply:
On the second thought, we cannot have a thread in shutdown while having other threads executing other methods of `GlobalHandler`. So, it should be OK, but I suggest removing lock_guards from this function.